src/VisualStudio/Core/Def/Implementation/ProjectSystem/RunningDocumentTableEventTracker.cs

```diff
@@ -0,0 +1,239 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.TextManager.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    /// <summary>
+    /// Class to register with the RDT and forward RDT events.
+    /// Handles common conditions before sending out notifications.
+    /// </summary>
+    internal class RunningDocumentTableEventTracker : IVsRunningDocTableEvents3, IDisposable
+    {
+        private readonly ForegroundThreadAffinitizedObject _foregroundAffinitization;
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+        private readonly IVsRunningDocumentTable4 _runningDocumentTable;
+        private uint _runningDocumentTableEventsCookie;
+
+        public event EventHandler<RunningDocumentTableEventArgs> OnCloseDocument;
+        public event EventHandler<RunningDocumentTableEventArgs> OnRefreshDocumentContext;
+        public event EventHandler<RunningDocumentTableEventArgs> OnReloadDocumentData;
+        public event EventHandler<RunningDocumentTableInitializedEventArgs> OnBeforeOpenDocument;
+        public event EventHandler<RunningDocumentTableInitializedEventArgs> OnInitializedDocument;
+        public event EventHandler<RunningDocumentTableRenamedEventArgs> OnRenameDocument;
```

Comment:
I wonder if it wouldn't be better to define an interface (e.g. `RunningDocumentTableEventListener`) that defines these as methods, instead of having 6 events and 3 classes that package arguments to these calls.

Reply:
Hmm and then the tracker would have the IRunningDocumentTableEventListener as a parameter and call into those methods instead of calling into the event?

Reply:
Just tried it out.  I can delete a bunch of code if I do it the interface way, so I think it's the way to go.

Reply:
should be much cleaner now, thanks for the suggestion :)